### PR TITLE
✨ Support for pod logs and other subresources

### PR DESCRIPTION
**Pull request by jmprusi**
## Summary

Adds support to kcp for POD subresources, adding the proper handler to use the Syncer Tunneler capabilities.

The Feature gate `KCPSyncerTunnel` should be enabled in KCP and all the desired Syncers:

```
--feature-gates KCPSyncerTunnel=true
``` 

There are still some missing parts to make this process "fully" automatic:

* Upsyncer controller and POD upsyncing.
* Automatically registering the POD resource.
* Sync command generating the YAML with the propper permissions for POD/$subresources to the Syncer `ClusterRole`.

### Steps to test this PR

Build the project and images:
```
make build build-kind-images-ko install
```

Run KCP:
```
./bin/kcp start --feature-gates KCPSyncerTunnel=true
```

Create your workspace:
```
KUBECONFIG=.kcp/admin.kubeconfig kubectl kcp ws create myorg --enter
KUBECONFIG=.kcp/admin.kubeconfig kubectl kcp ws create myworkspace --enter
```

Create your Synctarget, using the syncer imatge from the first step:
```
KUBECONFIG=.kcp/admin.kubeconfig kubectl kcp workload sync cluster1 --syncer-image=<syncer_image> -o syncer.yaml --resources=pods --feature-gates KCPSyncerTunnel=true
```

Now we need to modify the `syncer.yaml` file to add the missing `ClusterRole` permissions:

```diff
 ---
 apiVersion: rbac.authorization.k8s.io/v1
 kind: ClusterRole
 metadata:
   name: kcp-syncer-cluster2-33n2fkyd
 rules:
 - apiGroups:
   - ""
   resources:
   - namespaces
   verbs:
   - "create"
   - "list"
   - "watch"
   - "delete"
 - apiGroups:
   - "apiextensions.k8s.io"
   resources:
   - customresourcedefinitions
   verbs:
   - "get"
   - "watch"
   - "list"
 - apiGroups:
   - ""
   resources:
   - configmaps
   - secrets
   - services
   - pods
+  - pods/log
+  - pods/exec
   verbs:
   - "*"
 - apiGroups:
   - "apps"
   resources:
   - deployments
   verbs:
   - "*"
 - apiGroups:
   - "networking.k8s.io"
   resources:
   - ingresses
   verbs:
   - "*"
 ---
```
Deploy the syncer to out k8s cluster:
```
kubectl apply -f syncer.yaml
```

Let's bind the compute: 
```
KUBECONFIG=.kcp/admin.kubeconfig kubectl kcp bind compute root:myorg:myworkspace
```

We should create a deployment now:
```
KUBECONFIG=.kcp/admin.kubeconfig kubectl create deployment kuard-tunneler --image gcr.io/kuar-demo/kuard-amd64:blue
```

Now the tricky part, we need to create a POD in KCP with the same name as the downstream one trough the Upsyncer.

Get the upsyncer URL:
```
KUBECONFIG=.kcp/admin.kubeconfig kubectl get synctarget cluster1 --output=jsonpath={.status.virtualWorkspaces[0].url} | sed 's/syncer/upsyncer/'
```

The URL should look like: `https://172.22.10.60:6443/services/upsyncer/root:testing:testing/cluster1/ad76b43c-399b-40d7-b499-32a568e6861c`

Get the synctarget key:
```
KUBECONFIG=.kcp/admin.kubeconfig kubectl get synctarget cluster1 --output=jsonpath={.metadata.labels."internal\.workload\.kcp\.io\/key"}
```

Now we need the downstream POD name:
```
kubectl get pods --all-namespaces -l app=kuard-tunneler --output=jsonpath={.items..metadata.name}
```

Now that we have the Upsyncer VW URL, the Synctarget Key and the POD name we can upsync a "fake" POD (this will likely be done in the future by the upsyncer controller), please replace the <fields>:
```
cat <<EOF | KUBECONFIG=.kcp/admin.kubeconfig kubectl --server=<upsyncerVW url >/clusters/root:myorg:myworkspace apply -n default --validate=false -f -
apiVersion: v1
kind: Pod
metadata:
  name: <downstream-pod-name>
  namespace: default
  labels:
    app: kuard-tunneler
    state.workload.kcp.io/<synctargetkey>: Upsync
spec:
  containers:
    - name: kuard-amd64
EOF
```

If everything went Ok, you should be able to retrieve the POD in KCP:
```
$ KUBECONFIG=.kcp/admin.kubeconfig kubectl get pods
NAME                             READY   STATUS   RESTARTS   AGE
kuard-tunneler-f4bd64bd9-5p7jk   0/0              0          5s
```

Now you should be able to get the pods logs or exec a command:
```
$ KUBECONFIG=.kcp/admin.kubeconfig kubectl logs kuard-tunneler-f4bd64bd9-5p7jk
2022/11/25 17:33:41 Starting kuard version: v0.10.0-blue
2022/11/25 17:33:41 **********************************************************************
2022/11/25 17:33:41 * WARNING: This server may expose sensitive
2022/11/25 17:33:41 * and secret information. Be careful.
2022/11/25 17:33:41 **********************************************************************
2022/11/25 17:33:41 Config: 
{
  "address": ":8080",
  "debug": false,
  "debug-sitedata-dir": "./sitedata",
  "keygen": {
    "enable": false,
    "exit-code": 0,
    "exit-on-complete": false,
    "memq-queue": "",
    "memq-server": "",
    "num-to-gen": 0,
    "time-to-run": 0

...
```

## Related issue(s)

Fixes #https://github.com/kcp-dev/kcp/issues/1976

### CODE REVIEW

**Review by davidfestal — COMMENTED** [review]
Should we have unit tests ?

**Review by davidfestal — COMMENTED** [review]
Maybe move your handler in a dedicated file ? (like for the Home workspaces handler)

File: pkg/server/handler_test.go [hunk 1]
```diff
@@ -153,3 +164,104 @@ func TestProcessResourceIdentity(t *testing.T) {
...
+			}
+			r := httptest.NewRequest("GET", tc.path, nil).WithContext(request.WithCluster(ctx, request.Cluster{Name: logicalcluster.New(tc.workspace)}))
+			rw := httptest.NewRecorder()
+
+			handler := podSubresourceProxyHandlerBuilder.build()
+			handler.ServeHTTP(rw, r)
+			result := rw.Result()
+			bytes, err := io.ReadAll(result.Body)
+			require.NoError(t, err, "Request body cannot be read")
+			if tc.expectError {
```
Comment (davidfestal) [inline pkg/server/handler_test.go:258]: Don't you want to test the detail of the error, with something like `expectedError string` ?

Comment (jmprusi) [reply pkg/server/handler_test.go]: done

File: pkg/server/podsubresourceproxy_handler.go [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+	"github.com/kcp-dev/kcp/pkg/syncer/shared"
+	"github.com/kcp-dev/kcp/pkg/tunneler"
+)
+
+// WithPodSubresourceProxying proxies the POD subresource requests using the syncer tunneler.
+func WithPodSubresourceProxying(apiHandler http.Handler, kcpclient dynamic.ClusterInterface, kcpInformer kcpinformers.SharedInformerFactory) http.HandlerFunc {
+	r := regexp.MustCompile(`/api/v1/namespaces/(?P<Namespace>[^/]+)/pods/(?P<PodName>[^/]+)/(?P<Subresource>exec|log|portforward|proxy|attach)?(?P<Arguments>.*)`)
+	syncTargetInformer, err := kcpInformer.ForResource(schema.GroupVersionResource{Group: "workload.kcp.dev", Resource: "synctargets", Version: "v1alpha1"})
+	if err != nil {
+		panic(err)
```
Comment (davidfestal) [inline pkg/server/podsubresourceproxy_handler.go:51]: is it a common practice in the other handlers to `panic` ?

Comment (jmprusi) [reply pkg/server/podsubresourceproxy_handler.go:51]: yes, looks, panic is used in other handlers, for ex homeworkspace handler. But I can change it to return with a logger.error.

File: pkg/server/podsubresourceproxy_handler.go [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+				return nil, fmt.Errorf("expected 1 synctarget for key %q, got %d", synctargetKey, len(synctargets))
+			}
+			synctarget, ok := synctargets[0].(*workloadv1alpha1.SyncTarget)
+			if !ok {
+				return nil, fmt.Errorf("expected synctarget to be of type %T, got %T", &workloadv1alpha1.SyncTarget{}, synctargets[0])
+			}
+			return synctarget, nil
+		},
+	}
+	return builder.build()
```
[the remark this answers is not in the document]
Comment (jmprusi) [reply pkg/server/podsubresourceproxy_handler.go:84]: used this pattern to ease testing, I can change the method name to Handler() or so.

**Comment by ncdc** [comment]
/retitle ✨ Support for pod logs and other subresources

File: pkg/server/podsubresourceproxy_handler.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (davidfestal) [reply pkg/server/podsubresourceproxy_handler.go:84]: I don't see how it makes the test easier.
couldn't you directly assign the `podSubresourceProxyHandlerBuilder` struct (well a pointer to it probably) to the handler in your test implementation ?.

[the remark this answers is not in the document]
Comment (jmprusi) [reply pkg/server/podsubresourceproxy_handler.go]: done! thanks!

File: pkg/tunneler/tunnel.go [hunk 4]
```diff
@@ -96,6 +97,22 @@ func (rp *tunnelPool) deleteDialer(cluster, syncer string) {
 	delete(rp.pool, id)
 }
 
+// SyncerTunnelProxyPath builds the path for the proxy handler as expected by the Dialer
+func SyncerTunnelProxyPath(syncTargetWorkspaceName logicalcluster.Name, syncTargetName, downstreamNamespaceName, podName, subresource, arguments string) (url.URL, error) {
```
Comment (davidfestal) [inline pkg/tunneler/tunnel.go:101]: Is the function here used only once ?
Ten shouldn't it be private inside the use-site package ?

**Comment by jmprusi** [comment]
> I love seeing this progress ❤️
> 
> Am curious about the security story of this. Where do we do authorization?

@sttts the intention is to handle authorisation as part of the https://github.com/kcp-dev/kcp/issues/1972 epic, specifically issue: https://github.com/kcp-dev/kcp/issues/1974 but that would be on another PR(s) as a follow-up.

File: hack/logcheck.out [hunk 5]
```diff
@@ -127,16 +127,16 @@
...
-/pkg/tunneler/tunnel.go:146:3: function "V" should not be used, convert to contextual logging
-/pkg/tunneler/tunnel.go:178:5: function "Infof" should not be used, convert to contextual logging
-/pkg/tunneler/tunnel.go:178:5: function "V" should not be used, convert to contextual logging
-/pkg/tunneler/tunnel.go:182:4: function "Infof" should not be used, convert to contextual logging
-/pkg/tunneler/tunnel.go:182:4: function "V" should not be used, convert to contextual logging
-/pkg/tunneler/tunnel.go:195:4: function "Infof" should not be used, convert to contextual logging
-/pkg/tunneler/tunnel.go:195:4: function "V" should not be used, convert to contextual logging
-/pkg/tunneler/tunnel.go:230:4: function "Infof" should not be used, convert to contextual logging
-/pkg/tunneler/tunnel.go:230:4: function "V" should not be used, convert to contextual logging
+/pkg/tunneler/tunnel.go:147:3: function "InfoS" should not be used, convert to contextual logging
```
Comment (stevekuznetsov) [inline hack/logcheck.out:130]: /hold

At this point we're months in, and I think it's not appropriate to be deferring this any longer. You're touching the tunneler code, so please change these log lines to contextual logging in order to merge.

File: pkg/server/config.go [hunk 6]
```diff
@@ -368,6 +368,11 @@ func NewConfig(opts *kcpserveroptions.CompletedOptions) (*Config, error) {
 
 		if kcpfeatures.DefaultFeatureGate.Enabled(kcpfeatures.SyncerTunnel) {
 			apiHandler = tunneler.WithSyncerTunnel(apiHandler)
+			apiHandler = WithPodSubresourceProxying(
```
Comment (stevekuznetsov) [inline pkg/server/config.go:371]: @ncdc @sttts @davidfestal when TMC is "broken out" of the core API server, how does this work?

File: pkg/server/podsubresourceproxy_handler.go [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+	workloadv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/workload/v1alpha1"
+	kcpinformers "github.com/kcp-dev/kcp/pkg/client/informers/externalversions"
+	"github.com/kcp-dev/kcp/pkg/indexers"
+	"github.com/kcp-dev/kcp/pkg/syncer/shared"
+	"github.com/kcp-dev/kcp/pkg/tunneler"
+)
+
+// WithPodSubresourceProxying proxies the POD subresource requests using the syncer tunneler.
+func WithPodSubresourceProxying(apiHandler http.Handler, kcpclient dynamic.ClusterInterface, kcpInformer kcpinformers.SharedInformerFactory) http.Handler {
+	r := regexp.MustCompile(`/api/v1/namespaces/(?P<Namespace>[^/]+)/pods/(?P<PodName>[^/]+)/(?P<Subresource>exec|log|portforward|proxy|attach)?(?P<Arguments>.*)`)
```
Comment (stevekuznetsov) [inline pkg/server/podsubresourceproxy_handler.go:49]: Any reason we can't use the resource info structs from k8s and make assertions on those fields?

File: pkg/server/podsubresourceproxy_handler.go [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+		)
+		b.apiHandler.ServeHTTP(w, req)
+		return
+	}
+
+	// Let's get the synctargetKey
+	var synctargetKey string
+	for k, v := range pod.GetLabels() {
+		if strings.HasPrefix(k, workloadv1alpha1.ClusterResourceStateLabelPrefix) {
+			if v == "Upsync" {
```
Comment (stevekuznetsov) [inline pkg/server/podsubresourceproxy_handler.go:151]: No constant for this?

File: pkg/server/podsubresourceproxy_handler.go [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+			errorCodecs, schema.GroupVersion{}, w, req,
+		)
+		b.apiHandler.ServeHTTP(w, req)
+		return
+	}
+
+	synctarget, err := b.getSyncTargetBySynctargetKey(req.Context(), synctargetKey)
+	if apierrors.IsNotFound(err) {
+		responsewriters.ErrorNegotiated(
+			apierrors.NewNotFound(schema.GroupResource{Group: "workload.kcp.dev", Resource: "synctargets"}, synctargetKey),
```
Comment (stevekuznetsov) [inline pkg/server/podsubresourceproxy_handler.go:169]: Does this leak information the user would otherwise not be able to see?

File: pkg/server/podsubresourceproxy_handler.go [hunk 10]
```diff
@@ -0,0 +1,235 @@
...
+		responsewriters.ErrorNegotiated(
+			apierrors.NewInternalError(err),
+			errorCodecs, schema.GroupVersion{}, w, req,
+		)
+		b.apiHandler.ServeHTTP(w, req)
+		return
+	}
+
+	// Let's find the downstream namespace for the POD
+	// TODO(jmprusi): This is not enough, we should use a downstream Namespace informer with an index based on the namespaceLocator
```
Comment (stevekuznetsov) [inline pkg/server/podsubresourceproxy_handler.go:185]: TODO for when? Why not now?

Comment (stevekuznetsov) [reply pkg/server/podsubresourceproxy_handler.go:185]: Wouldn't this mean that the API server in `kcp` has an informer open against the pcluster?

File: pkg/server/config.go (hunk 6 above)
Comment (ncdc) [reply pkg/server/config.go:371]: Would we want to include this proxying in kcp core anyways, because it's "easier"?

Could/would we want to move this to the front proxy?

Comment (sttts) [reply pkg/server/config.go:371]: yes, front-proxy most likely

File: hack/logcheck.out (hunk 5 above)
Comment (jmprusi) [reply hack/logcheck.out]: Thanks, done!

File: pkg/server/podsubresourceproxy_handler.go (hunk 8 above)
Comment (jmprusi) [reply pkg/server/podsubresourceproxy_handler.go]: Done

File: pkg/server/podsubresourceproxy_handler.go (hunk 10 above)
Comment (jmprusi) [reply pkg/server/podsubresourceproxy_handler.go:194]: It needs discussion. I don't think we can have an informer open against the pcluster from KCP... perhaps some annotation on the upsynced POD to point to the correct downstream namespace?

File: pkg/server/podsubresourceproxy_handler.go (hunk 7 above)
Comment (jmprusi) [reply pkg/server/podsubresourceproxy_handler.go]: Fixed, thanks!

File: pkg/server/podsubresourceproxy_handler.go (hunk 9 above)
Comment (jmprusi) [reply pkg/server/podsubresourceproxy_handler.go]: Fixed, PTAL!

File: pkg/server/podsubresourceproxy_handler.go (hunk 10 above)
Comment (davidfestal) [reply pkg/server/podsubresourceproxy_handler.go:192]: Yeah, the same way we add `NamespaceLocator` annotations on synced downstream resources to remember the originating upstream namespace, the same way we could have an annotation on upsynced upstream resources, that would remember the originating downstream namespace.

Comment (davidfestal) [reply pkg/server/podsubresourceproxy_handler.go:192]: But I don't think it is urgent. Open an issue to not forget it should be sufficient for now.

Comment (davidfestal) [reply pkg/server/podsubresourceproxy_handler.go:192]: And maybe update the comment here, since it is currently quite confusing.

File: pkg/tunneler/tunnel.go [hunk 11]
```diff
@@ -32,12 +32,14 @@ import (
 	"k8s.io/klog/v2"
 
 	workloadv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/workload/v1alpha1"
+	"github.com/kcp-dev/logicalcluster/v3"
 )
 
 const (
-	defaultTunnelPathPrefix = "/services/syncer-tunnels/clusters"
+	// DefaultTunnelPathPrefix is the default path prefix for the tunnel service
+	DefaultTunnelPathPrefix = "/syncer-tunnels"
```
Comment (sttts) [inline pkg/tunneler/tunnel.go:40]: why was this moved outside of services? It was intentionally under `/services`.

Comment (davidfestal) [reply pkg/tunneler/tunnel.go:40]: @sttts was there a **strong** reason / hard requirement that made you use the `/services` URL ?

Comment (sttts) [reply pkg/tunneler/tunnel.go:40]: `/services` is for extensions. We will certainly not open up `/clusters/<sync target workspace>/<non-resource-url>`, i.e. add non-standard paths in the normal kube apiserver space.

Comment (sttts) [reply pkg/tunneler/tunnel.go:40]: We could maybe talk about a subresource of the synctarget.

Comment (davidfestal) [reply pkg/tunneler/tunnel.go:40]: Just a path change you mean ?
And we also have a filter on the the synctarget sub-resource, just like the one we have for the pod sub-resources ?

Comment (jmprusi) [reply pkg/tunneler/tunnel.go]: Done, now "connect" and "proxy" are subresources of the synctarget

**Comment by jmprusi** [comment]
Making sure I didn't introduce any flake.

/retest

File: pkg/tunneler/tunnel.go (hunk 11 above)
Comment (davidfestal) [reply pkg/tunneler/tunnel.go]: @sttts Is the implementation of this change (as done in commits 69ce74eee7b52fb4df79855e145fc101c32ac632 and 1c43ab5bcff9cd292346ba511a3cdb5247afffe0) OK for you ?

File: pkg/server/podsubresourceproxy_handler.go [hunk 12]
```diff
@@ -0,0 +1,245 @@
...
+	workloadv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/workload/v1alpha1"
+	kcpinformers "github.com/kcp-dev/kcp/pkg/client/informers/externalversions"
+	"github.com/kcp-dev/kcp/pkg/indexers"
+	"github.com/kcp-dev/kcp/pkg/syncer/shared"
+	"github.com/kcp-dev/kcp/pkg/tunneler"
+)
+
+// WithPodSubresourceProxying proxies the POD subresource requests using the syncer tunneler.
+func WithPodSubresourceProxying(apiHandler http.Handler, kcpclient dynamic.ClusterInterface, kcpInformer kcpinformers.SharedInformerFactory) http.Handler {
+	syncTargetInformer, err := kcpInformer.ForResource(schema.GroupVersionResource{Group: "workload.kcp.io", Resource: "synctargets", Version: "v1alpha1"})
```
Comment (davidfestal) [inline pkg/server/podsubresourceproxy_handler.go:47]:
```suggestion
	syncTargetInformer, err := kcpInformer.ForResource(workloadv1alpha1.SchemeGroupVersion.WithResource("synctargets"))
```

File: pkg/server/podsubresourceproxy_handler.go [hunk 13]
```diff
@@ -0,0 +1,245 @@
...
+	if err != nil {
+		panic(err)
+	}
+
+	return &podSubresourceProxyHandler{
+		apiHandler: apiHandler,
+
+		getPodByName: func(ctx context.Context, cluster logicalcluster.Name, namespace, podName string) (*corev1.Pod, error) {
+			unstr, err := kcpclient.Cluster(cluster.Path()).Resource(schema.GroupVersionResource{Group: "", Resource: "pods", Version: "v1"}).Namespace(namespace).Get(ctx, podName, metav1.GetOptions{})
```
Comment (davidfestal) [inline pkg/server/podsubresourceproxy_handler.go:56]:
```suggestion
			unstr, err := kcpclient.Cluster(cluster.Path()).Resource(corev1.SchemeGroupVersion.WithResource("pods")).Namespace(namespace).Get(ctx, podName, metav1.GetOptions{})
```
Or to avoid calling `Withresource` each time, you could create the `pods` GVR just above the `return`.

**Review by davidfestal — COMMENTED** [review]
Please have a look to @sttts commit-specific comments on commits 69ce74eee7b52fb4df79855e145fc101c32ac632 and 1c43ab5bcff9cd292346ba511a3cdb5247afffe0

File: pkg/server/podsubresourceproxy_handler.go (hunk 12 above)
Comment (jmprusi) [reply pkg/server/podsubresourceproxy_handler.go]: done

File: pkg/server/podsubresourceproxy_handler.go (hunk 13 above)
Comment (jmprusi) [reply pkg/server/podsubresourceproxy_handler.go]: done

File: pkg/server/podsubresourceproxy_handler.go [hunk 14]
```diff
@@ -123,7 +125,7 @@ func (b *podSubresourceProxyHandler) ServeHTTP(w http.ResponseWriter, req *http.
 		return
 	}
 
-	// Check if the subresource is valid, we only support exec, log, portforward, proxy and attach.
+	// Check if the subresource is valid, for PODs we support exec, log, portforward, proxy , attach, ephemeralcontainers.
 	if subresource != "exec" && subresource != "log" && subresource != "portforward" && subresource != "proxy" && subresource != "attach" {
```
Comment (davidfestal) [inline pkg/server/podsubresourceproxy_handler.go:129]: Where is the `ephemeralcontainers` in the `if` ?

File: pkg/server/podsubresourceproxy_handler_test.go [hunk 15]
```diff
@@ -163,7 +174,23 @@ func TestPodSubresourceProxyingHandler(t *testing.T) {
 				return
 			}
 			require.Equal(t, http.StatusOK, result.StatusCode, "Unexpected status code: %s", string(bytes))
-			require.Equal(t, tc.expectedPath, r.URL.Path, "Unexpected path")
+			if tc.expectedPath != "" {
```
Comment (davidfestal) [inline pkg/server/podsubresourceproxy_handler_test.go:177]: How is `expectedPath` useful now ? Why not remove it ?

File: pkg/tunneler/integration_test.go [hunk 16]
```diff
@@ -73,22 +77,25 @@ func setup(t *testing.T) (*http.Client, string, func()) {
 		publicServer.Close()
 		backend.Close()
 	}
-	return publicServer.Client(), dstURL + "/" + CmdTunnelProxy + "/", stop
+	return dstURL + "/" + proxySubresourcePath + "/", tunneler, stop
```
Comment (davidfestal) [inline pkg/tunneler/integration_test.go:80]: Do we really want the `proxy` subresource path here ?

File: pkg/tunneler/integration_test.go [hunk 17]
```diff
@@ -201,16 +211,19 @@ func Test_integration_listener_reconnect(t *testing.T) {
...
 	defer server2.Close()
 
-	resp, err = client.Get(uri) //nolint:noctx
-	if err != nil {
-		t.Fatalf("Request Failed: %s", err)
-	}
-	defer resp.Body.Close()
+	rw2 := httptest.NewRecorder()
+	b2 := &bytes.Buffer{}
+	req2, err := http.NewRequest(http.MethodGet, dstURL+"/"+proxySubresourcePath+"/", b2) //nolint:noctx
```
Comment (davidfestal) [inline pkg/tunneler/integration_test.go:216]: Why use the `proxySubresourcePath` here ? Does it still make sense ?

File: pkg/tunneler/tunnel.go [hunk 18]
```diff
@@ -36,8 +36,8 @@ import (
 )
 
 const (
-	cmdTunnelConnect = "connect"
-	CmdTunnelProxy   = "proxy"
+	tunnelSubresourcePath = "tunnel"
+	proxySubresourcePath  = "proxy"
```
Comment (davidfestal) [inline pkg/tunneler/tunnel.go:40]: Why do we still need this one ?

File: pkg/tunneler/tunnel.go [hunk 19]
```diff
@@ -120,27 +109,25 @@ func SyncerTunnelURL(host, ws, target string) (string, error) {
...
 }
 
 // WithSyncerTunnel adds an HTTP Handler that handles reverse connections and reverse proxy requests using 2 different paths exposing two synctarget subresources:
 //
-// https://host/clusters/<ws>/apis/workload.kcp.io/v1alpha1/synctargets/<name>/connect establish reverse connections and queue them so it can be consumed by the dialer
+// https://host/clusters/<ws>/apis/workload.kcp.io/v1alpha1/synctargets/<name>/tunnel establish reverse connections and queue them so it can be consumed by the dialer
 // https://host/clusters/<ws>/apis/workload.kcp.io/v1alpha1/synctargets/<name>/proxy/{path} proxies the {path} through the reverse connection identified by the cluster and syncer name
-func WithSyncerTunnel(apiHandler http.Handler) http.HandlerFunc {
-	pool := newTunnelPool()
+func (tn *Tunneler) SyncerTunnelHandler(apiHandler http.Handler) http.HandlerFunc {
```
Comment (davidfestal) [inline pkg/tunneler/tunnel.go:119]: Why rename the function ?

File: pkg/tunneler/tunnel.go [hunk 20]
```diff
@@ -120,27 +109,25 @@ func SyncerTunnelURL(host, ws, target string) (string, error) {
...
 	return func(w http.ResponseWriter, r *http.Request) {
 		logger := klog.FromContext(r.Context())
 
 		// fall through, syncer tunnels URL should contain the synctarget gvr.
-		if !strings.Contains(r.RequestURI, "workload.kcp.io/v1alpha1/synctargets") {
```
Comment (davidfestal) [inline pkg/tunneler/tunnel.go]: Can't we use `RequestInfo` to do those checks here now ?

File: pkg/server/config.go [hunk 21]
```diff
@@ -368,11 +368,13 @@ func NewConfig(opts *kcpserveroptions.CompletedOptions) (*Config, error) {
 		apiHandler = authorization.WithDeepSubjectAccessReview(apiHandler)
 
 		if kcpfeatures.DefaultFeatureGate.Enabled(kcpfeatures.SyncerTunnel) {
-			apiHandler = tunneler.WithSyncerTunnel(apiHandler)
+			tunneler := tunneler.NewTunneler()
+			apiHandler = tunneler.SyncerTunnelHandler(apiHandler)
 			apiHandler = WithPodSubresourceProxying(
```
Comment (davidfestal) [inline pkg/server/config.go:373]: Why couldn't `WithPodSubresourceProxying` be also a method of the `Tunneller` ?

File: pkg/tunneler/tunnel.go (hunk 20 above)
Comment (jmprusi) [reply pkg/tunneler/tunnel.go]: done

File: pkg/tunneler/tunnel.go (hunk 19 above)
Comment (jmprusi) [reply pkg/tunneler/tunnel.go]: done

File: pkg/tunneler/tunnel.go (hunk 18 above)
Comment (jmprusi) [reply pkg/tunneler/tunnel.go]: done

File: pkg/tunneler/integration_test.go (hunk 17 above)
Comment (jmprusi) [reply pkg/tunneler/integration_test.go]: done

File: pkg/server/podsubresourceproxy_handler_test.go (hunk 15 above)
Comment (jmprusi) [reply pkg/server/podsubresourceproxy_handler_test.go]: replaced by the expectedproxiedPath

File: pkg/server/podsubresourceproxy_handler.go (hunk 14 above)
Comment (jmprusi) [reply pkg/server/podsubresourceproxy_handler.go]: done

File: pkg/server/config.go (hunk 21 above)
Comment (jmprusi) [reply pkg/server/config.go]: Done

File: pkg/tunneler/tunnel.go [hunk 22]
```diff
@@ -105,138 +109,120 @@ func SyncerTunnelURL(host, ws, target string) (string, error) {
...
 			return
 		}
 
-		gv := workloadv1alpha1.SchemeGroupVersion
-		if path[1] != "apis" ||
-			path[2] != gv.Group ||
-			path[3] != gv.Version ||
-			path[4] != "synctargets" {
+		// Split the RequestURI to extract clusterName and SyncerName from the full path.
+		path := strings.Split(r.RequestURI, "/")
```
Comment (davidfestal) [inline pkg/tunneler/tunnel.go:138]: Not sure this Splitting is necessary. Don't you already have the information in the `RequestInfo.Parts` field (or possibly `RequestInfo.Path`  ?

Comment (davidfestal) [reply pkg/tunneler/tunnel.go:138]: There is also the `RequestInfo.Name`

Comment (jmprusi) [reply pkg/tunneler/tunnel.go]: Done!

File: pkg/server/podsubresourceproxy_handler_test.go (hunk 15 above)
Comment (jmprusi) [reply pkg/server/podsubresourceproxy_handler_test.go]: done

**Comment by davidfestal** [comment]
@sttts Would you give a last look and approve ?

**Comment by jmprusi** [comment]
known flake https://github.com/kcp-dev/kcp/issues/2389, retrying

/retest
